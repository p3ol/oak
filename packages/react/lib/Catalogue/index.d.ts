import { ComponentObject, ElementObject } from '@oakjs/core';
import { ReactNode, MutableRefObject, ComponentPropsWithRef } from 'react';

export declare type CatalogueRef = {
  open: () => void;
  close: () => void;
  toggle: () => void;
  opened: boolean;
  isOak: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface CatalogueProps extends ComponentPropsWithRef<any> {
  className?: string;
  placement?: string;
  onToggle?(props: { opened: boolean }): void;
  onAppend?(props: { component: ComponentObject }): void;
  onPaste?(clipboardData: ElementObject): void;
  ref?: MutableRefObject<CatalogueRef>;
}

declare function Catalogue(props: CatalogueProps): ReactNode | JSX.Element;

export default Catalogue;
