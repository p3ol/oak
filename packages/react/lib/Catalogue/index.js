import {
  forwardRef,
  useImperativeHandle,
  useReducer,
  useRef,
  useMemo,
} from 'react';
import { createPortal } from 'react-dom';
import {
  Tabs,
  Tab,
  classNames,
  mockState,
  ensureNode,
} from '@junipero/react';
import { slideInDownMenu } from '@junipero/transitions';
import {
  useFloating,
  useInteractions,
  useClick,
  useDismiss,
  offset,
  shift,
} from '@floating-ui/react';

import { useBuilder } from '../hooks';
import Icon from '../Icon';
import Text from '../Text';

const Catalogue = forwardRef(({
  component,
  className,
  placement = 'bottom',
  onToggle,
  onAppend,
  onPaste,
}, ref) => {
  const innerRef = useRef();
  const { builder, rootRef, rootBoundary, floatingsRef } = useBuilder();
  const [state, dispatch] = useReducer(mockState, {
    opened: false,
    clipboard: null,
  });
  const { x, y, refs, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: o => o ? open() : close(),
    middleware: [
      offset(16),
      shift({
        rootBoundary: rootBoundary?.current || rootRef?.current,
      }),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context),
    useDismiss(context),
  ]);

  useImperativeHandle(ref, () => ({
    open,
    close,
    toggle,
    opened: state.opened,
    isOak: true,
    innerRef,
  }));

  const open = () => {
    dispatch({ opened: true });
    onToggle?.({ opened: true });

    checkClipboard();
  };

  const checkClipboard = async () => {
    let clipboard;

    try {
      const element = JSON.parse(
        await globalThis.navigator.clipboard.readText()
      );

      if (Array.isArray(element)) {
        clipboard = element.filter(e => builder.getComponent(e.type));
      } else if (builder.getComponent(element.type)) {
        clipboard = element;
      }
    } catch (e) {}

    dispatch({ clipboard });
  };

  const close = () => {
    dispatch({ opened: false, clipboard: null });
    onToggle?.({ opened: false });
  };

  const toggle = e => {
    e.preventDefault();

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const onAppend_ = (component, e) => {
    e?.preventDefault();
    onAppend?.(component);
  };

  const availableGroups = builder.getAvailableComponents();
  const groups = useMemo(() => (
    availableGroups
      .filter(g => g.usable !== false)
      .map(g => ({
        ...g,
        components: g.components.filter(c =>
          c.usable !== false &&
          (!component || !component.disallow ||
            !component.disallow.includes(c.id))
        ),
      }))
      .filter(g => g.components.length)
  ), [availableGroups]);

  return (
    <div
      ref={innerRef}
      className={classNames(
        'catalogue',
        { opened: state.opened },
        className,
      )}
    >
      <a
        ref={refs.setReference}
        className={classNames(
          'handle oak-inline-flex oak-items-center',
          'oak-justify-center',
        )}
        draggable={false}
        { ...getReferenceProps() }
      >
        <Icon>add</Icon>
      </a>

      { floatingsRef?.current && createPortal(slideInDownMenu((
        <div
          ref={refs.setFloating}
          className="catalogue-menu"
          style={{
            position: strategy,
            top: y ?? 0,
            left: x ?? 0,
          }}
          data-placement={placement}
          { ...getFloatingProps() }
        >
          <div className="groups">
            <Tabs>
              { groups.map(group => (
                <Tab key={group.id} title={<Text>{ group.name }</Text>}>
                  <div className="group oak-grid oak-grid-cols-2 oak-gap-2">
                    { group.components.map(component => (
                      <a
                        key={component.id}
                        href="#"
                        draggable={false}
                        onClick={onAppend_.bind(null, component)}
                        className={classNames(
                          'component',
                          'component-' + component.id,
                          'oak-flex oak-items-center oak-px-2 oak-py-1',
                          'oak-gap-2 junipero'
                        )}
                      >
                        <Icon
                          className="!oak-text-2xl"
                          children={typeof component.icon === 'function'
                            ? component.icon.bind(null, component)
                            : component.icon}
                        />
                        <span className="name">
                          <Text>{ component.name }</Text>
                        </span>
                      </a>
                    )) }
                  </div>
                </Tab>
              )) }
            </Tabs>

            { state.clipboard && (
              <a
                onClick={onPaste?.bind(null, state.clipboard)}
                className={classNames(
                  'clipboard junipero oak-flex oak-items-center oak-px-2',
                  'oak-py-1 oak-gap-2',
                )}
              >
                <Icon>appearences</Icon>
                <Text name="core.pasteFromClipboard">
                  Paste from clipboard
                </Text>
              </a>
            ) }
          </div>
        </div>
      ), { opened: state.opened }), ensureNode(floatingsRef?.current)) }
    </div>
  );
});

Catalogue.displayName = 'Catalogue';

export default Catalogue;
