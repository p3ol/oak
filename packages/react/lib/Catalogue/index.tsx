import {
  type ComponentPropsWithRef,
  type MouseEvent,
  type MutableRefObject,
  forwardRef,
  useImperativeHandle,
  useMemo,
  useReducer,
  useRef,
} from 'react';
import type { ComponentObject, ElementObject } from '@oakjs/core';
import { createPortal } from 'react-dom';
import {
  type ForwardedProps,
  Tabs,
  Tab,
  classNames,
  mockState,
  ensureNode,
} from '@junipero/react';
import { slideInDownMenu } from '@junipero/transitions';
import {
  type UseFloatingOptions,
  type Boundary,
  useFloating,
  useInteractions,
  useClick,
  useDismiss,
  offset,
  shift,
  flip,
} from '@floating-ui/react';

import { useBuilder } from '../hooks';
import Icon from '../Icon';
import Text from '../Text';

export declare type CatalogueRef = {
  open: () => void;
  close: () => void;
  toggle: () => void;
  opened: boolean;
  isOak: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface CatalogueProps extends ComponentPropsWithRef<any> {
  className?: string;
  placement?: string;
  floatingOptions?: UseFloatingOptions & {
    boundary?: Boundary;
  };
  onToggle?(props: { opened: boolean }): void;
  onAppend?(component: ComponentObject): void;
  onPaste?(clipboardData: ElementObject): void;
  ref?: MutableRefObject<CatalogueRef>;
}

const Catalogue = forwardRef<CatalogueRef, CatalogueProps>(({
  component,
  className,
  placement = 'bottom',
  floatingOptions,
  onToggle,
  onAppend,
  onPaste,
}: CatalogueProps, ref) => {
  const innerRef = useRef();
  const { builder, rootRef, rootBoundary, floatingsRef } = useBuilder();
  const [state, dispatch] = useReducer(mockState, {
    opened: false,
    clipboard: null,
  });
  const { x, y, refs, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: o => o ? open() : close(),
    middleware: [
      offset(16),
      flip({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements.reference,
      }),
      shift({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements.reference,
      }),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context),
    useDismiss(context),
  ]);

  useImperativeHandle(ref, () => ({
    open,
    close,
    toggle,
    opened: state.opened,
    isOak: true,
    innerRef,
  }));

  const open = () => {
    dispatch({ opened: true });
    onToggle?.({ opened: true });

    checkClipboard();
  };

  const checkClipboard = async () => {
    let clipboard;

    try {
      const element = JSON.parse(
        await globalThis.navigator.clipboard.readText()
      );

      if (Array.isArray(element)) {
        clipboard = element.filter(e => builder.getComponent(e.type));
      } else if (builder.getComponent(element.type)) {
        clipboard = element;
      }
    } catch (e) {}

    dispatch({ clipboard });
  };

  const close = () => {
    dispatch({ opened: false, clipboard: null });
    onToggle?.({ opened: false });
  };

  const toggle = () => {
    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const onAppend_ = (
    component: ComponentObject,
    e: MouseEvent<HTMLAnchorElement>,
  ) => {
    e?.preventDefault();
    onAppend?.(component);
  };

  const availableGroups = builder.getAvailableComponents();
  const groups = useMemo(() => (
    availableGroups
      .filter(g => g.usable !== false)
      .map(g => ({
        ...g,
        components: g.components.filter((c: ComponentObject) =>
          c.usable !== false &&
          (!component || !component.disallow ||
            !component.disallow.includes(c.id))
        ),
      }))
      .filter(g => g.components.length)
  ), [availableGroups]);

  return (
    <div
      ref={innerRef}
      className={classNames(
        'catalogue',
        { opened: state.opened },
        className,
      )}
    >
      <a
        ref={refs.setReference}
        className={classNames(
          'handle oak-inline-flex oak-items-center',
          'oak-justify-center',
        )}
        draggable={false}
        { ...getReferenceProps() }
      >
        <Icon>add</Icon>
      </a>

      { floatingsRef?.current && createPortal(slideInDownMenu((
        <div
          ref={refs.setFloating}
          className="catalogue-menu"
          style={{
            position: strategy,
            top: y ?? 0,
            left: x ?? 0,
          }}
          data-placement={placement}
          { ...getFloatingProps() }
        >
          <div className="groups">
            <Tabs>
              { groups.map(group => (
                <Tab
                  key={group.id}
                  title={<Text>{ group.name }</Text>}
                >
                  <div className="group oak-grid oak-grid-cols-2 oak-gap-2">
                    { group.components.map((component: ComponentObject) => (
                      <a
                        key={component.id}
                        href="#"
                        draggable={false}
                        onClick={onAppend_.bind(null, component)}
                        className={classNames(
                          'component',
                          'component-' + component.id,
                          'oak-flex oak-items-center oak-px-2 oak-py-1',
                          'oak-gap-2 junipero'
                        )}
                      >
                        <Icon
                          className="!oak-text-2xl"
                          children={typeof component.icon === 'function'
                            ? component.icon.bind(null, component)
                            : component.icon}
                        />
                        <span className="name">
                          <Text>{ component.name as string }</Text>
                        </span>
                      </a>
                    )) }
                  </div>
                </Tab>
              )) }
            </Tabs>

            { state.clipboard && (
              <a
                onClick={onPaste?.bind(null, state.clipboard)}
                className={classNames(
                  'clipboard junipero oak-flex oak-items-center oak-px-2',
                  'oak-py-1 oak-gap-2',
                )}
              >
                <Icon>appearences</Icon>
                <Text name="core.pasteFromClipboard">
                  Paste from clipboard
                </Text>
              </a>
            ) }
          </div>
        </div>
      ),
      { opened: state.opened }), ensureNode(floatingsRef?.current)) }
    </div>
  );
});

Catalogue.displayName = 'Catalogue';

export default Catalogue;
