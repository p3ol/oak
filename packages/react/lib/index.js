export {
  TextField,
  SelectField,
  DateField,
  ColorField,
  Card,
  Dropdown,
  DropdownMenu,
  DropdownToggle,
  Toggle,
  TouchableZone,
  Label,
  classNames,
  useTimeout,
} from '@junipero/react';

export { default as Builder } from './Builder';
export { default as Container } from './Container';
export { default as Catalogue } from './Catalogue';
export { default as Icon } from './Icon';
export { default as Text } from './Text';
export { default as Field } from './Editable/Field';
export { default as DisplayableSettings } from './DisplayableSettings';

export {
  useBuilder,
  useLogger,
} from './hooks';

export * from './fields';
export * from './options';
export * from './addons';
