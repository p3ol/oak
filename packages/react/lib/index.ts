export {
  TextField,
  SelectField,
  DateField,
  ColorField,
  Card,
  Dropdown,
  DropdownMenu,
  DropdownToggle,
  Toggle,
  TouchableZone,
  Label,
  classNames,
  useTimeout,
} from '@junipero/react';

export {
  default as Builder,
  type BuilderProps,
} from './Builder';

export {
  default as Container,
  type ContainerProps,
} from './Container';

export {
  default as Catalogue,
  type CatalogueProps,
} from './Catalogue';

export {
  default as Icon,
} from './Icon';

export {
  default as Text,
  type TextProps as TextComponentProps,
} from './Text';

export {
  default as Field,
  type FieldProps,
} from './Editable/Field';

export {
  default as DisplayableSettings,
  type DisplayableSettingsProps,
} from './DisplayableSettings';

export {
  useBuilder,
  useLogger,
} from './hooks';

export * from './fields';
export * from './options';
export * from './addons';
