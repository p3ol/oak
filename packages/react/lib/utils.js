export const copyToClipboard = value =>
  globalThis.navigator.clipboard.writeText(value);

export const sanitizeHTML = content => {
  try {
    const parsed = new DOMParser().parseFromString(content, 'text/html');

    // Remove script & style tags
    parsed
      .querySelectorAll('script, style, iframe, object, video, audio')
      .forEach(item => item.parentNode.removeChild(item));

    // Disable all links
    parsed.querySelectorAll('a[href]').forEach(item => {
      item.removeAttribute('href');
      item.removeAttribute('target');
      item.setAttribute('role', 'link');
    });

    return parsed.body.innerHTML;
  } catch (e) {
    console.error(e);

    return '';
  }
};
