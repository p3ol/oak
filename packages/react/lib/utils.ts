export const copyToClipboard = (value: string): Promise<void> =>
  globalThis.navigator.clipboard.writeText(value);

export const sanitizeHTML = (content: string): string => {
  try {
    const parsed = new DOMParser().parseFromString(content, 'text/html');

    // Remove script & style tags
    parsed
      .querySelectorAll('script, style, iframe, object, video, audio')
      .forEach(item => item.parentNode.removeChild(item));

    // Disable all links
    parsed.querySelectorAll('a[href]').forEach(item => {
      item.removeAttribute('href');
      item.removeAttribute('target');
      item.setAttribute('role', 'link');
    });

    return parsed.body.innerHTML;
  } catch (e) {
    console.error(e);

    return '';
  }
};
