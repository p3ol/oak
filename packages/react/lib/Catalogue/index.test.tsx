import { fireEvent, render, screen } from '@testing-library/react';

import { BuilderLite } from '../../tests/utils';
import Catalogue from './index';

describe('<Catalogue />', () => {
  it('should allow to add elements to the builder', async () => {
    const onAppend = jest.fn();
    const { container, unmount } = render(
      <BuilderLite
        addons={[{
          components: [{
            id: 'test',
            group: 'core',
            type: 'component',
            name: 'Test component',
            construct: () => ({ type: 'test', content: 'Foo' }),
          }],
        }]}
      >
        <Catalogue onAppend={onAppend} />
      </BuilderLite>
    );

    expect(container).toMatchSnapshot('Closed');

    fireEvent.click(screen.getByText('add'));

    expect(container).toMatchSnapshot('Opened');

    fireEvent.click(screen.getByText('Test component'));

    expect(onAppend).toHaveBeenCalledTimes(1);

    unmount();
  });
});
