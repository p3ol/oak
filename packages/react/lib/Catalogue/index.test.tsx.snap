// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Catalogue /> should allow to add elements to the builder: Closed 1`] = `
<div>
  <div>
    <div
      class="catalogue"
    >
      <a
        class="handle oak-inline-flex oak-items-center oak-justify-center"
        draggable="false"
      >
        <i
          class="icon junipero-icons"
        >
          add
        </i>
      </a>
    </div>
  </div>
</div>
`;

exports[`<Catalogue /> should allow to add elements to the builder: Opened 1`] = `
<div>
  <div>
    <div
      class="catalogue opened"
    >
      <a
        class="handle oak-inline-flex oak-items-center oak-justify-center"
        draggable="false"
      >
        <i
          class="icon junipero-icons"
        >
          add
        </i>
      </a>
    </div>
    <div
      class="catalogue-menu"
      data-placement="bottom"
      style="position: absolute; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <div
        class="groups"
      >
        <div
          class="junipero tabs"
        >
          <ul
            class="titles"
          >
            <li
              class="title active"
            >
              <a
                href="#"
              >
                Other
              </a>
            </li>
          </ul>
          <div
            class="content"
          >
            <div
              class="tab"
            >
              <div
                class="group oak-grid oak-grid-cols-2 oak-gap-2"
              >
                <a
                  class="component component-test oak-flex oak-items-center oak-px-2 oak-py-1 oak-gap-2 junipero"
                  draggable="false"
                  href="#"
                >
                  <i
                    class="icon junipero-icons !oak-text-2xl"
                  />
                  <span
                    class="name"
                  >
                    Test component
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
